pinot-tools/src/main/java/com/linkedin/pinot/tools/realtime/provisioning/MemoryEstimator.java

```diff
@@ -0,0 +1,364 @@
+import com.linkedin.pinot.core.io.readerwriter.RealtimeIndexOffHeapMemoryManager;
+import com.linkedin.pinot.core.io.writer.impl.DirectMemoryManager;
+import com.linkedin.pinot.core.realtime.impl.RealtimeSegmentConfig;
+import com.linkedin.pinot.core.realtime.impl.RealtimeSegmentStatsHistory;
+import com.linkedin.pinot.core.segment.index.SegmentMetadataImpl;
+import java.io.File;
+import java.io.IOException;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.io.FileUtils;
+
+
+/**
+ * Given a sample segment, this class can estimate how much memory would be used per host, for various combinations of numHostsToProvision and numHoursToConsume
+ */
+public class MemoryEstimator {
+
+  private static final long MAX_MEMORY_BYTES = DataSize.toBytes("48G");
+  private static final String NOT_APPLICABLE = "NA";
+  private static final String TMP_DIR = System.getProperty("java.io.tmpdir") + File.separator;
+  private static final String STATS_FILE_NAME = "stats.ser";
+  private static final String STATS_FILE_COPY_NAME = "stats.copy.ser";
+
+  private File _sampleCompletedSegment;
+  private long _sampleSegmentConsumedSeconds;
+
+  private SegmentMetadataImpl _segmentMetadata;
+  private long _sampleCompletedSegmentSizeBytes;
+  private Set<String> _invertedIndexColumns;
+  private Set<String> _noDictionaryColumns;
+  int _avgMultiValues;
+  private File _tableDataDir;
+
+  private String[][] _totalMemoryPerHost;
+  private String[][] _optimalSegmentSize;
+  private String[][] _consumingMemoryPerHost;
+
+  public MemoryEstimator(File sampleCompletedSegment, long sampleSegmentConsumedSeconds) {
+    _sampleCompletedSegment = sampleCompletedSegment;
+    _sampleSegmentConsumedSeconds = sampleSegmentConsumedSeconds;
+
+    _sampleCompletedSegmentSizeBytes = FileUtils.sizeOfDirectory(_sampleCompletedSegment);
+    try {
+      _segmentMetadata = new SegmentMetadataImpl(_sampleCompletedSegment);
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception when reading segment index dir", e);
+    }
+
+    _noDictionaryColumns = _segmentMetadata.getAllColumns()
+        .stream()
+        .filter(column -> !_segmentMetadata.hasDictionary(column))
+        .collect(Collectors.toSet());
+    _invertedIndexColumns = _segmentMetadata.getAllColumns()
+        .stream()
+        .filter(column -> _segmentMetadata.getColumnMetadataFor(column).hasInvertedIndex())
+        .collect(Collectors.toSet());
+    _avgMultiValues = getAvgMultiValues();
+
+    _tableDataDir = new File(TMP_DIR, _segmentMetadata.getTableName());
+    try {
+      FileUtils.deleteDirectory(_tableDataDir);
+    } catch (IOException e) {
+      throw new RuntimeException("Exception in deleting directory " + _tableDataDir.getAbsolutePath(), e);
+    }
+    _tableDataDir.mkdir();
+  }
+
+  /**
+   * Initialize the stats file using the sample segment provided.
+   * This involves indexing each row of the sample segment using MutableSegmentImpl.
```

Comment:
you can add a comment here saying this is equivalent to consuming the rows of the segment, albeit in a different order than that consumed by the host. But the stats should be the same